### PR TITLE
Implementing JavaScript compiler (transpiler) and TypeScript support

### PR DESCRIPTION
**Pull request by fabiocav**

### CODE REVIEW

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+                var process = new Process { StartInfo = startInfo };
+                process.ErrorDataReceived += ProcessDataReceived;
+                process.OutputDataReceived += ProcessDataReceived;
+                process.EnableRaisingEvents = true;
+                process.Exited += (s, e) =>
+                {
+                    process.Close();
+
+                    // TODO: Fix, make sure output is read
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs]: Address this?

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                    Task.Delay(1000).ContinueWith(t =>
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs]: What is this magic delay, and what happens if it isn't long enough?

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            if (e.Data == null)
+            {
+                return;
+            }
+
+            Match match = DiagnosticRegex.Match(e.Data);
+
+            if (match.Success)
+            {
+                DiagnosticSeverity severity = (DiagnosticSeverity)Enum.Parse(typeof(DiagnosticSeverity), match.Groups[4].Value, true);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs]: Let's make sure we have test coverage for all this parsing logic

Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilation.cs]: Also, what if the version of the compiler changes and the output parsing you're doing fails to parse? How resilient are we?

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            };
+        }
+
+        public string Language => "TypeScript";
+
+        public IEnumerable<string> SupportedFileTypes => new[] { ".ts" };
+
+        private static string GetToolPath()
+        {
+            string path = ScriptSettingsManager.Instance.GetSetting(EnvironmentSettingNames.AzureWebJobsDisableHomepage);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs]: wrong setting

File: sample/HttpTrigger-TypeScript/index.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿export function index(context: any, req: any) {
```
Comment (mamaso) [inline sample/HttpTrigger-TypeScript/index.ts]: we'll need to add better type info for context and bindings - though probably out of scope for this PR.

**Comment by mathewc** [comment]
Recommend adding some more tests, e2e tests for typescript compilation, and some targeted tests for parsing logic, etc.

File: src/WebJobs.Script/Description/Node/Compilation/Raw/RawJavaScriptCompilationService.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        public string Language => "JavaScript";
+
+        public IEnumerable<string> SupportedFileTypes => new[] { ".js" };
+
+        async Task<object> ICompilationService.GetFunctionCompilationAsync(FunctionMetadata functionMetadata)
+            => await GetFunctionCompilationAsync(functionMetadata);
+
+        public Task<IJavaScriptCompilation> GetFunctionCompilationAsync(FunctionMetadata functionMetadata)
+        {
+            return _compilationResult;
```
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/Compilation/Raw/RawJavaScriptCompilationService.cs]: could use this for automatic npm install, though the perf implications might kill that.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 7]
```diff
@@ -17,23 +18,31 @@
 using Microsoft.Azure.WebJobs.Extensions.Http;
 using Microsoft.Azure.WebJobs.Host;
 using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Scripting;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 
+using ScriptFunc = System.Func<object, System.Threading.Tasks.Task<object>>;
```
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: that's a nice touch

File: sample/HttpTrigger-TypeScript/index.ts (hunk 5 above)
Comment (fabiocav) [inline sample/HttpTrigger-TypeScript/index.ts]: @christopheranderson is working on some of the type definitions, but the goal is to have at least the basic typings defined (e.g. context, request, response, etc.)

Comment (fabiocav) [inline sample/HttpTrigger-TypeScript/index.ts]: Even without those, we should be able to enhance this sample quite a bit. It's pretty plain right now.

File: src/WebJobs.Script/Description/Node/Compilation/Raw/RawJavaScriptCompilationService.cs (hunk 6 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/Node/Compilation/Raw/RawJavaScriptCompilationService.cs]: Yes. That thought crossed my mind. At least there's a clear hook, but we'll do that at a later time if desired.

Comment (christopheranderson) [inline src/WebJobs.Script/Description/Node/Compilation/Raw/RawJavaScriptCompilationService.cs]: Definitely a future thing. We might consider automatically "funcpack"-ing in the temp directory at some point which doesn't have the same file perf issues. Unrelated to TS support, though.

**Comment by christopheranderson** [comment]
Some additional things to test:
1. Bring in node packages and test that types resolve properly
2. Use ES6 features: Promise, object destructuring, async, generators 
3. Add scenarios for scriptFile/entrypoint

I've tested the following, which expresses (1) and some of (2) (Doesn't test generators, but async is a generator under the hood, so it's actually pretty close) and it works:

```typescript
import * as request from "request"

export function run(context, req) {
    context.log('1. Before promise');
    const p = new Promise((resolve, reject) => {
        context.log('3. Start request');
        request({
            uri: "https://jsonplaceholder.typicode.com/users",
            json: true
        }, (err, res, body) => {
            context.log('5. Resposne received');
            if(err || res.statusCode >= 400) reject(err || `Bad status code: ${res.statusCode}`)
            //call resolve instead of context.done();
            resolve(body);
        })
        context.log('4. After request');

        
    }).then((body) => {
        context.log('6a. Resposne from request');
        context.res = {
            // status: 200, /* Defaults to 200 */
            body: body
        }
    }).catch((err) => {
        context.log('6b. Resposne from request');
        context.res = {
            status: 500, /* Defaults to 200 */
            body: {
                err
            }
        }
    });
    context.log('2. After promise');
    return p;
};  
```

Output:
```
2017-04-20T22:07:42.312 Function started (Id=207937d9-7e8f-4556-8739-42acdc111104)
2017-04-20T22:07:48.933 1. Before promise
2017-04-20T22:07:48.933 3. Start request
2017-04-20T22:07:48.933 4. After request
2017-04-20T22:07:48.933 2. After promise
2017-04-20T22:07:48.996 5. Resposne received
2017-04-20T22:07:48.996 6a. Resposne from request
2017-04-20T22:07:49.197 Function completed (Success, Id=207937d9-7e8f-4556-8739-42acdc111104, Duration=6878ms)
```

**Comment by SteveALee** [comment]
wahoo! Thanks for this!

**Comment by SteveALee** [comment]
Out of interest,  where is the tsc config defined? Will you be exposing tsconfig.json - or at least a subset (things like outDir would be problematical).

I guess I should wait for the docs? :)

**Pull request merged**